src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -1167,17 +1140,196 @@ private boolean isSubtypeUncheckedInternal(Type t, Type s, boolean capture, Warn
+                        if (!isSameType(t2, s) &&
+                                t2.isReferenceProjection() == s.isReferenceProjection() &&
+                                structuralComparator.visit(t2, s)) {
+                            warn.warn(LintCategory.UNCHECKED);
+                            return true;
+                        }
+                        return false;
                     }
-                    return true;
                 }
             }
             return false;
         }
+        // where
+        StructuralTypeComparator structuralComparator = new StructuralTypeComparator();
+        class StructuralTypeComparator extends TypeRelation {
+            private Set<TypePair> cache = new HashSet<>();
+
+            @Override
+            public Boolean visitType(Type t, Type s) {
+                switch (t.getTag()) {
+                    case BYTE:
+                    case CHAR:
+                    case SHORT: case INT: case LONG:
+                    case FLOAT: case DOUBLE:
+                    case BOOLEAN: case VOID:
+                    case BOT: case NONE:
+                        return t.hasTag(s.getTag());
+                    case WILDCARD: //we shouldn't be here - avoids crash (see 7034495)
+                        return false;
+                    default:
+                        throw new AssertionError("isSubtype " + t.getTag());
+                }
+            }
+
+            @Override
+            public Boolean visitArrayType(ArrayType t, Type s) {
+                if (t == s)
+                    return true;
+
+                if (s.isPartial())
+                    return visit(s, t);
+
+                return s.hasTag(ARRAY)
+                        && visit(t.elemtype, elemtype(s));
+            }
+
+            @Override
+            public Boolean visitClassType(ClassType t, Type s) {
+                // If t is an intersection, sup might not be a class type
+                if (!t.hasTag(CLASS)) return visit(t, s);
+                return t.tsym == s.tsym
+                        && (t.tsym != s.tsym || t.referenceProjectionOrSelf().tsym == s.referenceProjectionOrSelf().tsym)
+                        && (!s.isParameterized() || compareTypeArgsRecursive(s, t))
+                        && visit(t.getEnclosingType(), s.getEnclosingType());
+            }
+
+            @Override
+            public Boolean visitTypeVar(TypeVar t, Type s) {
+                if (s.hasTag(TYPEVAR)) {
+                    TypeVar other = (TypeVar)s;
+                    if (allowUniversalTVars &&
+                            t.isValueProjection() != other.isValueProjection() && t.tsym == other.tsym) {
+                        return true;
+                    }
+                }
+                return isSameType(t, s);
+            }
+
+            public boolean compareTypeArgsRecursive(Type t, Type s) {
+                TypePair pair = new TypePair(t, s);
```

Comment:
This is typically needed in subtyping routines, to prevent against cases where we could expand terms forever. But in the case of a structural comparison, this can never happen?

Reply:
with wildcards yes